File: Desktop App/wpf/ArduinoAdmin/ArduinoAdmin/Lib/Message.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ArduinoAdmin.Lib
{
    class Message
    {
        private string buf;
        public Dictionary<string, string> Parameters{get;set;}
        private readonly char endOfParamName = '=';
        private readonly char endOfParamValue = ';';

        public enum MessageType
        {
            Command = 0,
            Configuration
        }

        public Message(MessageType type)
        {
            string strType;
            switch (type)
            {
                case MessageType.Command:
                    strType="CMD";
                    break;
                case MessageType.Configuration:
                    strType = "CFG";
                    break;
                default:
                    new ApplicationException("UnsupportedMessageType");
                    return;
            }
            this.buf = "MSG" + strType;
            //initialise le dictionnaire de paramètres
            this.Parameters = new Dictionary<string, string>();
        }

        public Message(string data)
        {
            this.buf = data.Substring(0,6);
            // eclate les arguments
            this.Parameters = data.Substring(6)
                .Split(';')
                .Select(part => part.Split(this.endOfParamName))
                .Where(part => part.Length == 2)
                .ToDictionary(sp => sp[0].ToUpper(), sp => sp[1]);
            Console.WriteLine(this.Parameters);
        }

        public string GetBuffer()
        {
            string strParam = String.Join(""+this.endOfParamValue, this.Parameters.Select(x => x.Key.ToUpper() + this.endOfParamName + x.Value).ToArray());
            return this.buf + strParam;
        }

        // ecrit la signature
        public bool EcrireSignature()
        {
            buf += "MSG";
            return true;
        }

        // lit la signature
        public int LireSignature(int ofs)
        {
            //lit la signature
            if (buf.Substring(ofs, 3) == "MSG")
                return ofs + 3;
            return 0;
        }

        // ecrit la signature
        public int EcrireParam(int ofs, string nom, string valeur)
        {
            string data = nom + this.endOfParamName + valeur + this.endOfParamValue;
            //ecrit les données
            buf += data;

            return ofs + data.Length;
        }

        // lit un parametre
        public int LireParam(int ofs, out string nom, out string valeur)
        {
            int endOfBuf = buf.Length;
            //lit le nom
            nom = "";
            while (ofs < endOfBuf && buf[ofs] != this.endOfParamName)
            {
                nom += buf[ofs];
            }
            ofs++;

            //lit la valeur
            valeur = "";
            while (ofs < endOfBuf && buf[ofs] != this.endOfParamValue)
            {
                valeur += buf[ofs];
            }
            ofs++;

            return ofs;
        } 
    }
}

File: Arduino App/Server/Message.cpp
#include "Message.h"

const char MessageTexte::endOfParamName = '=';
const char MessageTexte::endOfParamValue = ';';

//constructeur
MessageTexte::MessageTexte(){
  this->buf = NULL;
  this->endOfBuf = NULL;
}

MessageTexte::MessageTexte(char * buf, int sizeOfBuf){
  this->buf = buf;
  this->endOfBuf = buf + sizeOfBuf;
}

// obtient l'offset de base
char* MessageTexte::GetBuffer(){
  return this->buf;
}

// vérifie la structure du message
int MessageTexte::Verifier(){
        if(this->LireSignature(this->buf) == 0)
          return 0;
	return 1;
}

// lit le type de message
char* MessageTexte::LireType(char* ofs, int* type){
	if (ofs < buf || ofs + 3 > endOfBuf)
		return 0;

	if (ofs[0] == 'C' && ofs[1] == 'M' && ofs[2] == 'D'){
                *type = MessageTypeCommande;
		return ofs + 3;
        }
        
	if (ofs[0] == 'C' && ofs[1] == 'F' && ofs[2] == 'G'){
                *type = MessageTypeConfiguration;
		return ofs + 3;
        }
        
        return 0;
}

// ecrit la signature
char* MessageTexte::EcrireSignature(char* ofs){
	if (ofs < buf || ofs + 3 > endOfBuf)
		return 0;
	//lit la signature
	*ofs++ = 'M';
	*ofs++ = 'S';
	*ofs++ = 'G';
	return ofs;
}

// lit la signature
char* MessageTexte::LireSignature(char* ofs){
	if (ofs < buf || ofs > endOfBuf)
		return 0;
	//lit la signature
	if (ofs[0] == 'M' && ofs[1] == 'S' && ofs[2] == 'G')
		return ofs + 3;
	return 0;
}

// ecrit la signature
char* MessageTexte::EcrireParam(char* ofs, const char* nom, const char* valeur){
	int nomSize = strlen(nom);
	int valeurSize = strlen(valeur);

	if (ofs < buf || ofs > endOfBuf)
		return 0;

	if (nomSize < 1 || valeurSize < 1 || ofs + (nomSize + valeurSize + 2) >= endOfBuf)
		return 0;

	//ecrit les données
	memcpy(ofs, nom, nomSize);
	ofs += nomSize;

	*ofs++ = this->endOfParamName;

	memcpy(ofs, valeur, valeurSize);
	ofs += valeurSize;

	*ofs++ = this->endOfParamValue;

	return ofs;
}

// lit un parametre
char* MessageTexte::LireParam(char* ofs, char* nom, char* valeur){

	if (ofs < buf || ofs > endOfBuf)
		return 0;

	//lit le nom
	while (ofs < endOfBuf && *ofs != this->endOfParamName){
		*(nom++) = *(ofs++);
	}
	*nom = '\0';
	*ofs++;

	//lit la valeur
	while (ofs < endOfBuf && *ofs != this->endOfParamValue){
		*(valeur++) = *(ofs++);
	}
	*valeur = '\0';
	*ofs++;

	return ofs;
}
